test/e2e/cluster_logging_utils.go

```diff
@@ -19,51 +19,198 @@ package e2e
+	Name string
+	// If we didn't read some log entries, their
+	// timestamps should be no less than this timestamp.
+	// Effectively, timestamp of the last ingested entry
+	// for which there's no missing entry before it
+	LastTimestamp time.Time
+	// Cache of ingested and read entries
+	Occurrences map[int]*logEntry
+	// Number of lines expected to be ingested from this pod
+	ExpectedLinesNumber int
+}
+
+type logEntry struct {
+	Payload   string
+	Timestamp time.Time
+}
+
+type logsProvider interface {
+	EnsureWorking() error
+	ReadEntries(*loggingPod) []*logEntry
+}
+
+func (entry *logEntry) getLogEntryNumber() (int, bool) {
+	if len(entry.Payload) < logPrefixLength {
+		return 0, false
+	}
+	chunks := strings.Split(entry.Payload[logPrefixLength:], " ")
+	lineNumber, err := strconv.Atoi(strings.TrimSpace(chunks[0]))
+	return lineNumber, err == nil
+}
+
+func createLoggingPod(f *framework.Framework, podName string, totalLines int, loggingDuration time.Duration) *loggingPod {
+	framework.Logf("Starting pod %s", podName)
+	createLogsGeneratorPod(f, podName, totalLines, loggingDuration)
+
+	return &loggingPod{
+		Name: podName,
+		// It's used to avoid querying logs from before the pod was started
+		LastTimestamp:       time.Now(),
+		Occurrences:         make(map[int]*logEntry),
+		ExpectedLinesNumber: totalLines,
+	}
+}
+
+func createLogsGeneratorPod(f *framework.Framework, podName string, linesCount int, duration time.Duration) {
+	f.PodClient().Create(&api_v1.Pod{
+		ObjectMeta: meta_v1.ObjectMeta{
+			Name: podName,
 		},
-		Spec: v1.PodSpec{
-			RestartPolicy: v1.RestartPolicyOnFailure,
-			Containers: []v1.Container{
+		Spec: api_v1.PodSpec{
+			RestartPolicy: api_v1.RestartPolicyNever,
+			Containers: []api_v1.Container{
 				{
-					Name:  synthLoggerPodName,
-					Image: "gcr.io/google_containers/busybox:1.24",
-					// notice: the subshell syntax is escaped with `$$`
-					Command: []string{"/bin/sh", "-c", fmt.Sprintf("i=0; while [ $i -lt %d ]; do echo $i; i=`expr $i + 1`; done", linesCount)},
+					Name:  podName,
+					Image: "gcr.io/google_containers/logs-generator:v0.1.0",
+					Env: []api_v1.EnvVar{
+						{
+							Name:  "LOGS_GENERATOR_LINES_TOTAL",
+							Value: strconv.Itoa(linesCount),
+						},
+						{
+							Name:  "LOGS_GENERATOR_DURATION",
+							Value: duration.String(),
+						},
+					},
+					Resources: api_v1.ResourceRequirements{
+						Requests: api_v1.ResourceList{
+							api_v1.ResourceCPU: *resource.NewMilliQuantity(
+								loggingContainerCpu,
+								resource.DecimalSI),
+							api_v1.ResourceMemory: *resource.NewQuantity(
+								loggingContainerMemory,
+								resource.BinarySI),
+						},
+					},
 				},
 			},
 		},
 	})
 }
 
-func reportLogsFromFluentdPod(f *framework.Framework) error {
-	synthLoggerPod, err := f.PodClient().Get(synthLoggerPodName, metav1.GetOptions{})
+func waitForLogsIngestion(logsProvider logsProvider, pods []*loggingPod, ingestionTimeout time.Duration) error {
```

Comment:
You export hundreds of thousands of lines in the load test and expect all of them to be correctly ingested here while we don't guarantee 100% delivery rate. Please add some logic here to allow some failure rate. We should start far below our target to avoid flakiness and then harden it over time up to our target.